src/solvers/petsc_dm_wrapper.C

```diff
@@ -16,39 +16,355 @@
 // Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 
 #include "libmesh/libmesh_common.h"
+#include "libmesh/petsc_macro.h"
 
 #ifdef LIBMESH_HAVE_PETSC
+#if !PETSC_VERSION_LESS_THAN(3,7,3)
 
 #include "libmesh/ignore_warnings.h"
 #include <petscsf.h>
+#include <petsc/private/dmimpl.h>
```

Comment:
It's usually inadvisable to include or rely on `petsc/private` headers or any of their internal implementation stuff. But this may be unavoidable if you are using relatively new features that aren't otherwise accessible...

Reply:
In this case, it's currently unavoidable because we need this function and PETSc put it as private. Will lobby @knepley next week to make it public (or tell us what the better thing to do is).